### PR TITLE
chore: drop 3.6 support

### PR DESCRIPTION
**Pull request by agoose77**
- Replace text references to 3.6 with 3.7 where appropriate
- Remove CI 3.6 special-cases
- Replace manylinux1 with manylinux2014

### CODE REVIEW

**Comment by agoose77** [comment]
@jpivarski I've taken the liberty of removing manylinux1 support at the same time, as [we only need it for 3.6 as far as I can tell](https://labs.quansight.org/blog/2020/11/manylinux1-is-obsolete-manylinux2010-is-almost-eol-what-is-next). However, I noticed that our CI was building manylinux1 wheels for 3.6, 3.7, and 3.8. Do you know why that is? It makes me hesitant to remove it.

**Comment by jpivarski** [comment]
> @jpivarski I've taken the liberty of removing manylinux1 support at the same time, as [we only need it for 3.6 as far as I can tell](https://labs.quansight.org/blog/2020/11/manylinux1-is-obsolete-manylinux2010-is-almost-eol-what-is-next). However, I noticed that our CI was building manylinux1 wheels for 3.6, 3.7, and 3.8. Do you know why that is? It makes me hesitant to remove it.

We should probably get a sign-off from @henryiii before merging this (between 1.9.0 and 1.10.0, following https://github.com/scikit-hep/awkward/issues/1626#issuecomment-1227458731). The time it takes to release 1.9.0rc13 and 1.9.0 (two releases == 2 days) might correspond to the time when he's back and ready to review things.

**Comment by agoose77** [comment]
@jpivarski I want to (at some point) rename the CI workflows (both filename and display names):
- `test-wheel-construction.yml` (`Wheels`)
- `test-compilation.yml` (`Build`)
- `deploy-wheels` (`Deploy`)

Maybe we want to move the linting into a new workflow?
- `test-linters.yml` (`Lint`)

E.g.
image


Is this PR as good a time as any to do this?
In any case, we already need to fiddle with the required CI checks due to this PR (incidentally, I think my changes to the CI need to be further considered - the non-3.6 Linux jobs no longer match those in the expected checks)

**Comment by jpivarski** [comment]
Yes, this is as good a time as any. By doing it before the git split (`main` and `main-v1`), the names of these tests can be the same in both branches without having to manually make both branches agree.

**Comment by jpivarski** [comment]
It reminds me that I'll have to make two branch protection rules. (I've already done that for Uproot; just forgot about it.) In principle, the test names can differ because the two branches can have different rules, but it will be a lot easier to keep track of if they're not different. (Otherwise, we'd be saying, "main test X corresponds to main-v1 test Y...")

**Comment by henryiii** [comment]
There are a very small number of users on 3.7 & 3.8 that might have a default pip that can only access manylinux1. They would already be likely to be having some issues, and more in the future. They are also using a version of pip that does not officially support Python 3.7 and 3.8 (it predated those versions) but is forced onto those Pythons (and possibly patched) by linux distro managers.

We should also be bumping pyupgrade to 3.7+, and enabling `from __future__ import annotations` injection by isort. Static typing is much easer in Python 3.7! (wink, wink)

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -61,8 +61,8 @@ jobs:
     - uses: pypa/cibuildwheel@v2.9.0
       env:
         CIBW_ARCHS_MACOS: universal2
-        CIBW_MANYLINUX_X86_64_IMAGE: manylinux1
-        CIBW_BUILD: cp39-win_amd64 cp36-manylinux_x86_64 cp38-macosx_universal2
+        CIBW_MANYLINUX_X86_64_IMAGE: manylinux2014
```
Comment (henryiii) [inline .github/workflows/tests.yml]: This can go away now, it can (and is, I think) be set in pyproject.toml.

File: .github/workflows/wheels.yml [hunk 2]
```diff
@@ -62,8 +56,8 @@ jobs:
     - uses: pypa/cibuildwheel@v2.9.0
       env:
         CIBW_BUILD: ${{ matrix.build }}
-        CIBW_MANYLINUX_X86_64_IMAGE: ${{ matrix.CIBW_MANYLINUX_X86_64_IMAGE }}
         CIBW_ARCHS: ${{ matrix.arch }}
+        CIBW_MANYLINUX_X86_64_IMAGE: manylinux2014
```
Comment (henryiii) [inline .github/workflows/wheels.yml]: Same.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -267,11 +267,11 @@ Templating is only used for integer specialization.
...
+We target Python 3.7 and above. Import statements can assume Python 3 names, string-checking can assume Python 3 meanings of `str` and `bytes`, Unicode literals don't need to be prefixed by `u`, and dict order can be assumed to be stable. Python's f-strings can now be used, but not with equals signs (e.g. `f"{something = }"` rather than `f"something = {something}"`) because that's a Python 3.8 feature (its main use is in debugging, anyway).
 
-If you see any outdated (pre-Python 3.6) code, you can safely clean them up. Some strings are not easier to read as f-strings or require some work to make them more readable; it should be a case-by-case basis.
+If you see any outdated (pre-Python 3.7) code, you can safely clean them up. Some strings are not easier to read as f-strings or require some work to make them more readable; it should be a case-by-case basis.
 
-Python 3.6 will be the standard until it becomes "painful" to keep it (like, when CI and build tools no longer support it, or if we adopt type annotations).
+Python 3.7 will be the standard until it becomes "painful" to keep it (like, when CI and build tools no longer support it, or if we adopt type annotations).
```
Comment (henryiii) [inline CONTRIBUTING.md]: Is this valid, or are we beginning to follow [official CPython EoL](https://endoflife.date/python)? (Assuming not NEP 29 / SPEC 1, as that's 3.8+ currently 😉)

File: README-pypi.md [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 [![PyPI version](https://badge.fury.io/py/awkward.svg)](https://pypi.org/project/awkward)
 [![Conda-Forge](https://img.shields.io/conda/vn/conda-forge/awkward)](https://github.com/conda-forge/awkward-feedstock)
-[![Python 3.6‒3.10](https://img.shields.io/badge/python-3.6%E2%80%923.10-blue)](https://www.python.org)
+[![Python 3.7‒3.10](https://img.shields.io/badge/python-3.7%E2%80%923.10-blue)](https://www.python.org)
```
Comment (henryiii) [inline README-pypi.md]: 3.7-3.11?

Comment (jpivarski) [reply README-pypi.md]: You're right that it should go up to 3.11, thanks to the tests that you added.

File: CONTRIBUTING.md (hunk 3 above)
Comment (jpivarski) [reply CONTRIBUTING.md]: Yeah, let's go with the CPython EoL dates. I am changing my mind from when I wrote that (from "maximum comfortable support" to "let's just follow a schedule"). I didn't have a particular schedule in mind, but the EoL dates sounds like a very defensible policy.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (agoose77) [reply .github/workflows/tests.yml]: Oh yes, I forgot that we use `pyproject.toml` as a configuration source even in the CI.

File: CONTRIBUTING.md (hunk 3 above)
Comment (agoose77) [reply CONTRIBUTING.md]: Is the new wording appropriate? I.e., are we proposing to drop Python versions alongside CPython's EoL schedule?

Comment (agoose77) [reply CONTRIBUTING.md]: > We should also be bumping pyupgrade to 3.7+, and enabling `from __future__ import annotations` injection by isort. Static typing is much easer in Python 3.7! (wink, wink)

I'm stoked about this!

**Comment by jpivarski** [comment]
I'm going to make a 1.10.0rc1 out of this. The only changes I made were to CONTRIBUTING.md, so I'm not going to wait for the tests to finish.

If we need to make any more fixes, it will be 1.10.0rc2.

Thanks!!!

**Pull request merged**